File: code/cpp/topsort.cpp
#include <iostream>
#include <stack>
#include <vector>
#include <string>
#include <algorithm>
#include "helpers.h"

int dfs(int i, int node, std::vector<bool> &visited, std::vector<int> &visitedNodes, networks::Graph &graph);
std::vector<int> topsort(networks::Graph &graph);

int main()
{   
#ifndef FILE_INPUT
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif

    int numNodes, numEdges;
    std::cin >> numNodes >> numEdges;

    bool directed, weighted;
    std::cin >> directed >> weighted;

    std::vector<networks::Edge> edges;
    if (weighted)
    {
        for (int i=0; i<numEdges; i++)
        {
            int u, v, w;
            std::cin >> u >> v >> w;
            edges.push_back(networks::Edge(u, v, w));
        }
    }
    else
    {
        for (int i=0; i<numEdges; i++)
            {
                int u, v;
                std::cin >> u >> v;
                edges.push_back(networks::Edge(u, v));
            }
    }
    
    networks::Graph graph = networks::Graph(numNodes, edges, directed);
    std::vector<int> ordering;

    ordering = topsort(graph);
    helpers::printVector(ordering);

    return 0;
}


std::vector<int> topsort(networks::Graph &graph)
{
    int n = graph.numNodes;
    std::vector<bool> visited(n, false);
    std::vector<int> ordering(n);
    int i= n-1;

    for (int node=0; node<n; node++)
    {
        if (!visited[node])
            i = dfs(i, node, visited, ordering, graph);
        
    }

    return ordering;    
}

int dfs(int i, int node, std::vector<bool> &visited, std::vector<int> &ordering, networks::Graph &graph)
{
    visited[node] = true;

    std::vector<std::pair<int, int>> edges = graph.adjList[node];
    for (auto edge : edges)
    {
        if (!visited[edge.first])
            i = dfs(i, edge.first, visited, ordering, graph);
    }

    ordering[i] = node;
    return i-1;
}

File: code/python/topsort.py
import helpers as H


# global variables
INPUT_PATH = '/home/freyr/dev/GraphTheory/code/python/input.txt'
OUTPUT_PATH = '/home/freyr/dev/GraphTheory/code/python/output.txt'

# clearing the contents of the file if any
open(OUTPUT_PATH, 'w').close()


def topsort(graph: H.Graph):
    n = graph.num_nodes
    visited = [False for _ in range(n)]
    ordering = [0 for _ in range(n)]
    i = n-1
    
    for node in range(n):
        if not visited[node]:
            wrapper = [visited]  # making a wrapper to pass by reference
            i = dfs(i, node, wrapper, ordering, graph)
            
    return ordering
    
            
def dfs(i, node, v_wrapper, ordering, graph: H.Graph):
    v_wrapper[0][node] = True
    edges = graph.adj_list[node]
    for edge in edges:
        if not v_wrapper[0][edge[0]]:
             i = dfs(i, edge[0], v_wrapper, ordering, graph)
             
    ordering[i] = node
    return i-1

if __name__ == '__main__':
    with open(INPUT_PATH, 'r') as f:
        num_nodes, num_edges = (int(i) for i in f.readline().split(" "))
        directed, weighted = (int(i) for i in f.readline().split(" "))
        
        edges = []
        if weighted:
            for i in range(num_edges):
                u, v, w = (int(i) for i in f.readline().split(" "))
                edges.append(H.Edge(u, v, w))
        else:
            for i in range(num_edges):
                u, v = (int(i) for i in f.readline().split(" "))
                edges.append(H.Edge(u, v))
                
    graph = H.Graph(num_nodes, edges, directed)
    
    ordering = topsort(graph)
    with open(OUTPUT_PATH, 'a') as f:
        f.write(f"{ordering}")